src/config/validation.rs

```diff
@@ -189,6 +194,192 @@ pub fn check_outputs(config: &ConfigBuilder) -> Result<(), Vec<String>> {
+                // On the other hand, we need to be sure this code works correctly on all of our
+                // "tier 1" platforms -- i.e. Linux, Windows, macOS, and in containerized
+                // environments -- before we could feel confident enough, I think, to block Vector
+                // startup on such a validation.
+                //
+                // If we can indeed show that this code works correctly on those environments, such
+                // that we can be reasonably confident that normal users won't encounter errors that
+                // hamstring them, then maybe we could make this an actual failure instead of just a
+                // warning log.
+                None => warn!(
+                    buffer_id = usage.id().id(),
+                    buffer_data_dir = usage.data_dir().to_string_lossy().as_ref(),
+                    resolved_buffer_data_dir = resolved_data_dir.to_string_lossy().as_ref(),
+                    message = "Found no matching mountpoint for buffer data directory.",
+                ),
+                Some(mountpoint) => {
+                    let entries = mappings.entry(mountpoint.clone()).or_insert_with(Vec::new);
+
+                    entries.push(usage);
+                }
+            }
+
+            mappings
+        });
+
+    // Finally, we have a mapping of disk buffers, based on their underlying mountpoint. Go through
+    // and check to make sure the sum total of `max_size` for all buffers associated with each
+    // mountpoint does not exceed that mountpoint's total capacity.
+    //
+    // We specifically do not do any sort of warning on free space because that has to be the
+    // responsibility of the operator to ensure there's enough total space for all buffers present.
+    let mut errors = Vec::new();
+
+    for (mountpoint, buffers) in mountpoint_buffer_mapping {
+        let buffer_max_size_total: u64 = buffers.iter().map(|usage| usage.max_size()).sum();
+        let mountpoint_total_capacity = mountpoints
+            .get(&mountpoint)
+            .copied()
+            .expect("mountpoint must exist");
+
+        if buffer_max_size_total > mountpoint_total_capacity {
+            let component_ids = buffers
+                .iter()
+                .map(|usage| usage.id().id())
+                .collect::<Vec<_>>();
+            errors.push(format!(
+                "Mountpoint '{}' has total capacity of {} bytes, but configured buffers using mountpoint have total maximum size of {} bytes. \
+Reduce the `max_size` of the buffers to fit within the total capacity of the mountpoint. (components associated with mountpoint: {})",
+                mountpoint.to_string_lossy(), mountpoint_total_capacity, buffer_max_size_total, component_ids.join(", "),
+            ));
+        }
+    }
+
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(errors)
+    }
+}
+
+#[cfg(target_os = "macos")]
+fn get_resolved_dir_path(path: &Path) -> PathBuf {
```

Comment:
I'd note that this setup is similar to some immutable style Linux distro's - not that we explicitly support them today, but we may want to expand that support. Another thought as I type this, when running in Kubernetes we frequently/by default set the root as read only.

Would definitely be good to check everything works as expected in k8s.